akka-stream/src/main/scala/akka/stream/impl/Sinks.scala

```diff
@@ -456,110 +449,130 @@ import scala.collection.generic.CanBuildFrom
 /**
  * INTERNAL API
  */
-@InternalApi final private[stream] class LazySink[T, M](sinkFactory: T ⇒ Future[Sink[T, M]], zeroMat: () ⇒ M) extends GraphStageWithMaterializedValue[SinkShape[T], Future[M]] {
+@InternalApi final private[stream] class LazySink[T, M](sinkFactory: T ⇒ Future[Sink[T, M]]) extends GraphStageWithMaterializedValue[SinkShape[T], Future[Option[M]]] {
   val in = Inlet[T]("lazySink.in")
   override def initialAttributes = DefaultAttributes.lazySink
   override val shape: SinkShape[T] = SinkShape.of(in)
 
   override def toString: String = "LazySink"
 
   override def createLogicAndMaterializedValue(inheritedAttributes: Attributes) = {
-    lazy val decider = inheritedAttributes.mandatoryAttribute[SupervisionStrategy].decider
 
-    var completed = false
-    val promise = Promise[M]()
+    val promise = Promise[Option[M]]()
     val stageLogic = new GraphStageLogic(shape) with InHandler {
+      var switching = false
       override def preStart(): Unit = pull(in)
 
       override def onPush(): Unit = {
+        val element = grab(in)
+        switching = true
+        val cb: AsyncCallback[Try[Sink[T, M]]] =
+          getAsyncCallback {
+            case Success(sink) ⇒
+              // check if the stage is still in need for the lazy sink
+              // (there could have been an onUpstreamFailure in the meantime that has completed the promise)
+              if (!promise.isCompleted) {
```

Comment:
FYI not a good pattern to guard like this; in some cases can be racy between the isCompleted check and the success call. Usually `trySuccess` is the way to go. In this case it'll be fine since only one trigger for `cb` and concurrency guarded due to the stage and callback, but tip in general